cdap-app-fabric/src/main/java/co/cask/cdap/app/runtime/ProgramRuntimeProvider.java

```diff
@@ -39,7 +39,12 @@
     /**
      * Returns the list of supported {@link ProgramType}.
      */
-    ProgramType[] value();
+    ProgramType[] types();
+
+    /**
+     * Returns the Spark Compat version if Spark is a supported {@link ProgramType}.
+     */
+    SparkCompat sparkCompat();
```

Comment:
The `ProgramRuntimeProvider` should be generic that can support any type of program runtime (ideally we should have MR runner implementing this an in a separate extension).

In order to support compat module, which could be the need for different program runtime as well, such as MR 2.3 vs MR 2.7, we could have `AbstractExtensionLoader.getAllExtensions` to call a method on the `EXTENSION` instead of based on the result of `getSupportedTypesForProvider` call only (it requires some refactoring of the `AbstractExtensionLoader` and `ProgramRuntimeProvider` class, but shouldn't be too big).

By doing so, the `SparkProgramRuntimeProvider` could depends on the environment (in distributed mode) or `cConf` setting (in SDK) to say if it can provide a program runtime or not.